src/v/storage/tests/index_state_test.cc

```diff
@@ -97,3 +98,126 @@ BOOST_AUTO_TEST_CASE(encode_decode_future_version) {
     auto dst = storage::index_state::hydrate_from_buffer(src_buf.copy());
     BOOST_REQUIRE(!dst);
 }
+
+// encode/decode using new serde framework
+BOOST_AUTO_TEST_CASE(serde_basic) {
+    for (int i = 0; i < 100; ++i) {
+        auto input = make_random_index_state();
+        const auto input_copy = input.copy();
+        BOOST_REQUIRE_EQUAL(input, input_copy);
+
+        // objects are equal
+        const auto buf = serde::to_iobuf(std::move(input));
+        auto output = serde::from_iobuf<storage::index_state>(buf.copy());
+        BOOST_REQUIRE_EQUAL(output, input_copy);
+
+        // round trip back to equal iobufs
+        const auto buf2 = serde::to_iobuf(std::move(output));
+        BOOST_REQUIRE_EQUAL(buf, buf2);
+    }
+}
+
+// accept decoding supported old version
+BOOST_AUTO_TEST_CASE(serde_supported_deprecated) {
+    for (int i = 0; i < 100; ++i) {
+        auto input = make_random_index_state();
+        const auto output = serde::from_iobuf<storage::index_state>(
+          storage::serde_compat::index_state_serde::encode(input));
```

Comment:
This is a good point.

In the case of index_state I've done a manual test involving a simulated upgrade using an older version of redpanda. The stakes aren't super high for index_state because any failure (e.g. size or crc mismatch) would result in the index being transparently rebuilt. But... this is the only case in the tree where that is true.

We don't have a huge amount of data that is serialized right now, which makes me think that it is a good time to build out some tooling for managing / verifying a collection of blobs before it gets out of hand. Mind adding any thoughts you have on this to this issue: https://github.com/vectorizedio/redpanda/issues/3200 and I'll prioritize this before the next PR that swaps in the serde format.